### ISSUE TITLE
The reason of using r_inter_tanh_s8 = torch.tanh(r_inter_s8)

### ISSUE THREAD

**Issue by lfxx** [hkchengrex/CascadePSP#49]
I read your code,in pspnet.py,line107:
```
r_inter_tanh_s8 = torch.tanh(r_inter_s8)
```
I can't find any explainments about this in your paper.Want to know why we need torch.tanh instead of relu or sigmoid?

**Comment by lfxx** [comment]
resloved

**Comment by hkchengrex** [comment]
See https://github.com/hkchengrex/CascadePSP/issues/2

**Comment by lfxx** [comment]
> See #2

Get it.Thanks!Another question,RM module:
![image](https://user-images.githubusercontent.com/27956643/123944923-345f0700-d9d0-11eb-8ae4-4a5f1295b339.png)
We can see the second,third up structures don't have shortcut,but in your code,pspnet.py:
https://github.com/hkchengrex/CascadePSP/blob/a7f7836275d6fa914d320d0bfbe6c19cfb6c2f8d/models/psp/pspnet.py#L53-L63
there also have a shortcut regardless of upsample1,2,3.This's quite confusing.According to your paper,upsample2,upsample3 ops should not have shortcut.

**Comment by hkchengrex** [comment]
That's just part of the UP block...

**Comment by lfxx** [comment]
> That's just part of the UP block...

So is there a full version of photo that can denote the specific procedure of RM module?Really confused...
I modify the photo above,could you please ensure if it is right?
![111](https://user-images.githubusercontent.com/27956643/123968150-911aeb80-d9e9-11eb-9e9a-448dc0bc9cae.png)

**Comment by hkchengrex** [comment]
No it's not.
You are probably referring to the global step, in which the RM (in the figure) is used. The connections that you added kind of exist in the global step, but not in a single RM module. Note that the RM module is used thrice in the global step.
I was referring to the residual connection which you cited in code.

**Comment by lfxx** [comment]
> No it's not.
> You are probably referring to the global step, in which the RM (in the figure) is used. The connections that you added kind of exist in the global step, but not in a single RM module. Note that the RM module is used thrice in the global step.
> I was referring to the residual connection which you cited in code.

Thanks.May i have your training log as i get a shock loss when i use this repo to train.The training datasets are as same as you which contains MSRA-10K,ECSSD,etc..the hyper parameters are same as this repo too.The difference from this repo is:
i add se_attention_module and dcnv2 module to resnet50.
The code is running on a single RTX3090 using pytorch1.9 with batch_size 12.the loss graph is below:
![image](https://user-images.githubusercontent.com/27956643/124446772-4fa48a80-ddb3-11eb-90e0-bc7ed84b0acf.png)
As you can see,up to 1.4k iter_numbers,the loss is always up and down around 0.4.How should i debug this?Seems like decrease lr is a method.
upate:up to 3k iter_numbers,the loss is decreasing very slowly.

**Comment by hkchengrex** [comment]
Are the visual results fine?
Decreasing LR might be a way.

**Comment by lfxx** [comment]
> Are the visual results fine?
> Decreasing LR might be a way.

After 47500 iters,the model is not converge yet.
GT is:
![gt](https://user-images.githubusercontent.com/27956643/124533744-7f9c6e00-de45-11eb-9a93-4b393a4a0920.png)
OS1 is:
![os1](https://user-images.githubusercontent.com/27956643/124533778-8f1bb700-de45-11eb-9261-dc97a2938934.png)
Seems like os1 is more poor than gt.Is os1 good enough?Look forward to your kindly reply and give me further advice!

**Comment by hkchengrex** [comment]
They look fine in general. You might try simple grid-search / simply train longer to see if you can get a better result. The loss might not be too indicative of the final performance since it lacks the global/local cascade.

**Comment by lfxx** [comment]
> They look fine in general. You might try simple grid-search / simply train longer to see if you can get a better result. The loss might not be too indicative of the final performance since it lacks the global/local cascade.

Thanks.In your code,layer3 and layer4 has dilation=2 and 4.layer3 has  continous 6 bottleneck and layer4 has 3.Why the result doesn't have gridding effect?
Now i want to replace resnet50 with mobilenetv3,latter has some 5×5 kernel size while former only has 3x3 kernel size,in those bottlenecks,do i need to use dilation conv to enlarge Receptive Field?

**Comment by hkchengrex** [comment]
Gridding is not inevitable, esp. with binary output.
It should be a good idea to make the receptive field the same/similar.

**Comment by lfxx** [comment]
> Gridding is not inevitable, esp. with binary output.
> It should be a good idea to make the receptive field the same/similar.

The result the repo outputs is quite good without gridding effect,in mobilenetv3,what should i do to avoid gridding effect?
Is that means for kernel 3,we need dilation=4;for kernel 5,we need dilation=2?
Another question,you return  x_1 without relu,but x_2 and x is relued.what's the difference between them?

**Comment by hkchengrex** [comment]
1. We just didn't have gridding (maybe cascade helps to mitigate). Nothing special was done.
2. I'm not familiar with the particulars of MobileNet, so you might have to figure that out yourself.
3. Those are very minor differences and should not have any significant effects.

**Comment by lfxx** [comment]
> 1. We just didn't have gridding (maybe cascade helps to mitigate). Nothing special was done.
> 2. I'm not familiar with the particulars of MobileNet, so you might have to figure that out yourself.
> 3. Those are very minor differences and should not have any significant effects.

Very detained explaination.Thanks!

**Comment by lfxx** [comment]
update:
Change resnet50 to mobilenetv3_large,
with fp32,On BIG_PSPNET_SS,i get 0.02% IOU decrease,0.4% mBA decrease and 1.4X speed up!
with fp16,On BIG_PSPNET_SS,the acc is same as fp32 but get 2.1X speed up compared to resnet50!
